### ISSUE TITLE
mm/alloc: undefined behavior in `SvsmAllocator::dealloc()`

### ISSUE THREAD

**Issue by 00xc** [coconut-svsm/svsm#107]
[Miri](/rust-lang/miri) reports undefined behavior when running the `mm::alloc::test_slab_alloc_free_many` test:

```console
$ MIRIFLAGS="-Zmiri-permissive-provenance" cargo +nightly miri test --target=x86_64-unknown-linux-gnu test_slab_alloc_free_many
Preparing a sysroot for Miri (target: x86_64-unknown-linux-gnu)... done
    Finished test [unoptimized + debuginfo] target(s) in 0.01s
     Running unittests src/lib.rs (target/miri/x86_64-unknown-linux-gnu/debug/deps/svsm-4dadd712fd0a3e32)

running 1 test
test mm::alloc::test_slab_alloc_free_many ... error: Undefined Behavior: trying to retag from <wildcard> for SharedReadWrite permission at alloc174[0x0], but no exposed tags have suitable permission in the borrow stack for this location
    --> src/mm/alloc.rs:1276:17
     |
1276 |                 (*slab).deallocate(virt_addr);
     |                 ^^^^^^^
     |                 |
     |                 trying to retag from <wildcard> for SharedReadWrite permission at alloc174[0x0], but no exposed tags have suitable permission in the borrow stack for this location
     |                 this error occurs as part of two-phase retag at alloc174[0x0..0x40]
     |
     = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
     = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
     = note: BACKTRACE:
     = note: inside `<mm::alloc::SvsmAllocator as core::alloc::GlobalAlloc>::dealloc` at src/mm/alloc.rs:1276:17: 1276:24
note: inside `mm::alloc::test_slab_alloc_free_many`
    --> src/mm/alloc.rs:1569:26
     |
1569 |                 unsafe { ALLOCATOR.dealloc(*p, layout) };
     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure
    --> src/mm/alloc.rs:1539:32
     |
1537 | #[test]
     | ------- in this procedural macro expansion
1538 | // Allocate and free a couple of objects for each slab size.
1539 | fn test_slab_alloc_free_many() {
     |                                ^
     = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

error: test failed, to rerun pass `--lib`

Caused by:
  process didn't exit successfully: `/home/carlos/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo-miri runner /home/carlos/dev/svsm-gh/target/miri/x86_64-unknown-linux-gnu/debug/deps/svsm-4dadd712fd0a3e32 test_slab_alloc_free_many` (exit status: 1)
note: test exited abnormally; to see the full output pass --nocapture to the harness.
```

**Comment by 00xc** [comment]
My guess is this stems from not going through any `Slab` lock whatsoever when freeing a pointer, since we access the `Slab` by casting a `VirtAddr` to a raw pointer:

https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L1268-L1280

**Comment by joergroedel** [comment]
Right, not taking any lock here is an actual bug. I suggest we fix it by not storing a pointer to the slab object in PageInfo, but the allocation size it is used for. From there we can get to the real slab object and take the lock for it.

**Comment by 00xc** [comment]
We don't actually need to store the size since we get it via the `layout` parameter, right? Or do we want to store it for another purpose?

**Comment by joergroedel** [comment]
Only if we trust the layout we get from the caller 😊 But we also need it to make a check whether we free to a normal allocation slab and the SLAB_PAGE_SLAB.

**Comment by Freax13** [comment]
`dealloc` is unsafe. Its [safety section](https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html#safety-2) states that the layout must be the same that was passed in when the pointer was allocated. For those reasons you can trust the layout parameter.

**Comment by 00xc** [comment]
Moreover, if I understand correctly, most of the time the layout should come from the Rust compiler, which will insert calls to `dealloc()` when values are dropped. I think we can make the `ALLOCATOR` global not `pub`, avoiding manual calls altogether.

> But we also need it to make a check whether we free to a normal allocation slab and the SLAB_PAGE_SLAB.

But we already know this through the page info, right? e.g. either `Page::Allocated` or `Page::SlabPage`. Or am I mixing things up?

**Comment by joergroedel** [comment]
> Moreover, if I understand correctly, most of the time the layout should come from the Rust compiler, which will insert calls to `dealloc()` when values are dropped. I think we can make the `ALLOCATOR` global not `pub`, avoiding manual calls altogether.

Okay, let me look into using the layout directly.

> But we already know this through the page info, right? e.g. either `Page::Allocated` or `Page::SlabPage`. Or am I mixing things up?

No, the difference between Page::Allocated and Page::SlabPage is the allocation size. For any allocation over 2KiB it will be Page::Allocated.

**Comment by 00xc** [comment]
I opened a PR to partially fix this, but there are more UB warnings down the call path of `Slab::shrink_slab()`.

**Comment by Freax13** [comment]
I investigated the other instance of UB:
https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L1075-L1090
If the first page (i.e. `self.common.page`) is free, we remove it with `self.common.remove_slab_page(unsafe { &mut *last_page }, slab_page); `. The problem with this is that we created two mutable references to `self.common.page`, one through `last_page` and another implicitly through `self.common`. This is obviously not allowed and it seems like the `as *mut SlabPage` was added to stop the borrow checker from complaining about this.
